src/cryptonote_core/blockchain.h

```diff
@@ -112,6 +113,93 @@ namespace cryptonote
       uint64_t already_generated_coins; //!< the total coins minted after that block
     };
 
+    /**
+     * @brief serialize the access to blockchain data
+    */
+    struct blockchain_transaction {
+
+      blockchain_transaction() : r_threads(0), w_locks(0)
+      {};
+
+      void start_read() {
+        while (!all_locks_released()) {
+          w_condition.wait(w_mutex);
+        }
+        {
+          boost::lock_guard<decltype(_g_mutex)> lock(_g_mutex);
+          r_mutex.lock();
+          one_more_read_thread();
+        }
+      };
+
+      void start_write() {
+        while (!is_zero_thread_reading()) {
+          r_condition.wait(r_mutex);
+        }
+        {
+          boost::lock_guard<decltype(_g_mutex)> lock(_g_mutex);
+          w_mutex.lock();
+          one_more_write_lock();
+        }
+      };
+
+      void end_read() {
+        boost::lock_guard<decltype(_g_mutex)> lock(_g_mutex);
+        one_read_thread_finished();
+        if(is_zero_thread_reading()) {
+          r_mutex.unlock();
+          r_condition.notify_all();
+        }
+      };
+
+      void end_write() {
+        boost::lock_guard<decltype(_g_mutex)> lock(_g_mutex);
+        w_mutex.unlock();
+        one_write_lock_released();
+        if(all_locks_released()) {
+          w_condition.notify_all();
+        }
+      };
+
+      ~blockchain_transaction() { };
+
+    private:
+
+      inline void one_more_read_thread() {
+        r_threads.fetch_add(1, std::memory_order::memory_order_seq_cst);        
+      }
+
+      inline void one_read_thread_finished() {
+        r_threads.fetch_sub(1, std::memory_order::memory_order_seq_cst);
+      }
+
+      inline bool is_zero_thread_reading() { 
+        return !r_threads.load(std::memory_order::memory_order_seq_cst);
+      }
+
+      inline void one_more_write_lock() {
+        w_locks.fetch_add(1, std::memory_order::memory_order_seq_cst);
+      }
+
+      inline void one_write_lock_released() {
+        w_locks.fetch_sub(1, std::memory_order::memory_order_seq_cst);        
+      }
+
+      inline bool all_locks_released() {
+        return !w_locks.load(std::memory_order::memory_order_seq_cst);
+      }
+
+      boost::mutex _g_mutex;
+
+      boost::shared_mutex r_mutex;
+      std::atomic<unsigned int> r_threads;  
+      boost::condition_variable_any r_condition;
+
+      boost::recursive_mutex w_mutex;
+      std::atomic<unsigned int> w_locks;
```

Comment:
Why would there allowed to be more than 1 write lock at a time?

Reply:
How much would the code have to be refactored so that we don't use the locks recursively in the first place? (probably a lot)

Reply:
That is one of the cleaner approaches and end result of that approach would be much much cleaner. But the problem with that is a lot of APIs should change.

Reply:
take a look at this line: 

https://github.com/0xFFFC0000/monero/blob/0574c09993f32c3ec532b9511b57fccb0818f2bc/src/cryptonote_core/blockchain.cpp#L119

This is much cleaner. But at the moment, not possible. Since we have to change a lot of APIs.